perl_genrule.bzl

```diff
@@ -0,0 +1,125 @@
+"""Generate files with perl. These are assumed to be .pl files as src and .s file as output.
+"""
+
+load("@rules_cc//cc:defs.bzl", "CcInfo", "cc_common")
+
+def combine_list_of_lists(list_of_lists):
+    final_list = []
+    for lst in list_of_lists:
+        final_list = final_list + lst
+    return final_list
+
+def generate_single_command(binary, assembly_flavor, src, out, ctx):
+    """Find the sources and outs and the command for a single src and out.
+
+    Args:
+        binary: The binary to run
+        assembly_flavor: The assembly flavor to produce
+        src: The source to use
+        out: The output to produce
+        ctx: The bazel rule context
+    Returns:
+        A list with the command to run, The source files and the out files.
+    """
+    out_files = []
+    src_files = []
+    commands = []
+    out_file = ctx.actions.declare_file(out)
+    src_files = src.files.to_list()
+
+    # We only care about the first source since there should only be
+    src_file = src_files[0]
+    command = "{} {} {} {}".format(binary, src_file.path, assembly_flavor, out_file.path)
+    commands.append(command)
+    src_files.append(src_file)
+    out_files.append(out_file)
+    return commands, src_files, out_files
+
+def generate_commands(binary, assembly_flavor, srcs_to_outs, srcs_to_outs_dupes, ctx):
+    """Generate commands needed to produces outs from sources. 
+
+    Args:
+        binary: The binary to run
+        assembly_flavor: The type of assembly to produce
+        srcs_to_outs: The main sources to outputs dict
+        srcs_to_outs_dupes: The secondary sources to outputs dict
+        ctx: The bazel rule context
+    Returns:
+        The commands joined on comma, the source files and the output files
+    """
+
+    commands = []
+    out_files = []
+    src_files = []
+    for src, out in srcs_to_outs.items():
+        intermediate_commands, intermediate_src_files, intermediate_out_files = generate_single_command(binary, assembly_flavor, src, out, ctx)
+        commands = commands + intermediate_commands
+        out_files = out_files + intermediate_out_files
+        src_files = src_files + intermediate_src_files
+    for src, out in srcs_to_outs_dupes.items():
+        intermediate_commands, intermediate_src_files, intermediate_out_files = generate_single_command(binary, assembly_flavor, src, out, ctx)
+        commands = commands + intermediate_commands
+        out_files = out_files + intermediate_out_files
+        src_files = src_files + intermediate_src_files
+    if ctx.attr.is_unix:
+        return ",".join(commands), src_files, out_files
+    else:
+        return ";".join(commands), src_files, out_files
+
+def _perl_genrule_impl(ctx):
+    binary_invocation = "perl"
```

Comment:
This adds a runtime dependency on a system-installed `perl` - can we use `perl` from `rules_perl` when not on Windows instead?

Reply:
Roughly:
1. Add a `rules_perl` dependency to the `MODULE.bazel`
2. Add `config_common.toolchain_type("@rules_perl//perl:current_toolchain", mandatory=False)` to the `toolchains` attribute on the rule
3. In the rule implementation, access `ctx.toolchains["@rules_perl//perl:current_toolchain"]` - it will be `None` on Windows, or a target on non-Windows
4. You can add `toolchain.files` as inputs, and access either `toolchain.perl_runtime.interpreter` (which should be either a `File` or `Path`), or use the make variable `$(PERL)` and expand it with https://bazel.build/rules/lib/builtins/ctx#expand_make_variables